### ISSUE TITLE
configuring snap version via john.conf

### ISSUE THREAD

**Issue by mac-arno** [openwall/john-packages#49]
There is apparently no documentation about how John can be configured in the Snap version. The John snap filesystem is intentionally write protected and therefore the /snap/johne-the-ripper/current/run/john.conf cannot be altered. Where in the userspace is John looking for an alternate john.conf? This question should be discussed in the Snap version docs.

**Comment by claudioandre-br** [comment]
Thanks for reporting. 

It's actually a bit confusing, but we just have to think about how the upstream itself works:

1. you cannot use `john-local.conf` (at least I haven't found a way to use it since `john` doesn't have a built-in variable that points to your runtime `Private home`);
2. so, you must have a full `john.conf` in the snap (environment) proper HOME folder. It is `$HOME/snap/john-the-ripper`;
3. Probably would be nice if upstream `john` has a variable that points to the `Private home`.

So let's see how it works using an example:

DO NOT REMOVE YOUR `john.pot` IF IT MATTERS TO YOU. Save it first, at least.

****

* I'm followiing the edge snap channel. Works the same way if you are using stable.
```shell
$ cd ~

$ john --list=build-info
Version: 1.9.0-jumbo-1+bleeding-3553509ce9 2023-05-08 13:39:36 +0200
Build: linux-gnu 64-bit x86_64 AVX2 AC OMP OPENCL
SIMD: AVX2, interleaving: MD4:3 MD5:3 SHA1:1 SHA256:1 SHA512:1
Deploy: sandboxed as a Snap app
System-wide exec: /snap/john-the-ripper/current/run
System-wide home: /snap/john-the-ripper/current/run
Private home: ~/.john
[...]
$JOHN is /snap/john-the-ripper/current/run/
[...]
```

```
# Create some sample hashes
john --format=sha512crypt -list=format-tests | cut -f3 > ~/alltests.in
```

* Crack it using the "regular" packaged `john`
```bash
$ rm -f snap/john-the-ripper/current/.john/*.pot; john ~/alltests.in --max-run=10 --mask
Warning: detected hash type "sha512crypt", but the string is also recognized as "sha512crypt-opencl"
Use the "--format=sha512crypt-opencl" option to force loading these as that type instead
Using default input encoding: UTF-8
Loaded 6 password hashes with 4 different salts (1.5x same-salt boost) (sha512crypt, crypt(3) $6$ [SHA512 256/256 AVX2 4x])
Cost 1 (iteration count) is 5000 for all loaded hashes
Will run 8 OpenMP threads
Using default mask: ?1?2?2?2?2?2?2?3?3?3?3?d?d?d?d    ## <===== HERE ************************************
Press 'q' or Ctrl-C to abort, 'h' for help, almost any other key for status
0g 0:00:00:00  (2) 0g/s 0p/s 0c/s 0C/s
                 (?)     
1g 0:00:00:02 2.49% (3) (ETA: 12:08:28) 0g/s 728.8p/s 2550c/s 4008C/s e5..5ya
1g 0:00:00:10 8.70% (3) (ETA: 12:09:01) 0g/s 702.1p/s 2306c/s 3710C/s Eli..lzi
Use the "--show" option to display all of the cracked passwords reliably
Session stopped (max run-time reached)

```

* Use your personal (edited) `john.conf` (compare the mask).
```bash
$ cp /snap/john-the-ripper/current/run/john.conf snap/john-the-ripper/current/.john/john.conf

$ sed -i 's/DefaultMask = ?1?2?2?2?2?2?2?3?3?3?3?d?d?d?d/DefaultMask = Hello?awor?l?l?a/g' snap/john-the-ripper/current/.john/john.conf

$ rm -f snap/john-the-ripper/current/.john/*.pot; john ~/alltests.in --max-run=10 --mask
Warning: detected hash type "sha512crypt", but the string is also recognized as "sha512crypt-opencl"
Use the "--format=sha512crypt-opencl" option to force loading these as that type instead
Using default input encoding: UTF-8
Loaded 6 password hashes with 4 different salts (1.5x same-salt boost) (sha512crypt, crypt(3) $6$ [SHA512 256/256 AVX2 4x])
Cost 1 (iteration count) is 5000 for all loaded hashes
Will run 8 OpenMP threads
Using default mask: Hello?awor?l?l?a   ## <===== HERE ************************************
Press 'q' or Ctrl-C to abort, 'h' for help, almost any other key for status
0g 0:00:00:00  (10) 0g/s 0p/s 0c/s 0C/s
                 (?)     
1g 0:00:00:02 3.09% (11) (ETA: 12:09:26) 0g/s 726.2p/s 2541c/s 3994C/s Hello.worg..Hello7woria
1g 0:00:00:10 10.82% (11) (ETA: 12:09:54) 0g/s 700.7p/s 2302c/s 3703C/s HelloIworje..Hello9worni
Use the "--show" option to display all of the cracked passwords reliably
Session stopped (max run-time reached)
```

**Comment by claudioandre-br** [comment]
Lets think about the issue as documentation. I don't have a better place to add this note for now.

**Comment by claudioandre-br** [comment]
Actually, the question and the answer apply the same way to, for example, the Docker image. You will need add a:
```
-v "$(pwd)":/home/JtR
```

To connect the 'host' file system to the running container (linking the host current folder to `/home/JtR`). 

Then create a suitable `$(pwd)/.john/john.conf` file in your 'host'.

****
Hey, why is `john.conf` inside the `.john` folder? I don't care, anywhere upstream thinks it is valid is ok with me.

**Comment by claudioandre-br** [comment]
Moved to the wiki: [Configuring packages via john.conf](https://github.com/openwall/john-packages/wiki/Configuring-packages-via-john.conf)

**Comment by mac-arno** [comment]
Here are my own findings following yours for an approach to solve the problem:
According your approach I got `command not found` when calling `$ john --list=build-info`. So I discovered two installed versions of John on my Debian system. To be shure to use the right - Snap installed - version of John I changed into the Snap directory of John with
`$ cd /snap/john-the-ripper/current/run` and run it with
`$ ./john --list=build-info` That worked. This version was asking for missing config and log files in `/home/<user>/.john/`. Following this I put an altered copy of john.conf into this directory. When starting John again via 
`$ ./john` it uses the config-file put in my user directory as described above throwing out.

**Comment by mac-arno** [comment]
No, there are no problems with John in Debian at all, not from my experience. But I have installed two versions: the apt version and the Snap version. The apt version is a little bit cut off concerning the options. Therefore I installed the Snap version. And by calling John via „john“ alias I got started the apt version. That’s the difference I wanted to explain for other users. Thanks for let me know that ‚john-the-ripper‘ calls the Snap John.
